docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1060 @@
+- The node generation is also useful for managing remote objects which are not per-tenant,
+  such as the persistent per-node deletion queue which is proposed in this RFC.
+- node generation numbers enable building fencing into any network protocol (for example
+  communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
+
+Node that the two generation numbers have a different behavior for stale generations:
+
+- A pageserver with a stale generation number should immediately terminate: it is never intended
+  to have two pageservers with the same node ID.
+- An attachment with a stale generation number is permitted to continue operating (ingesting WAL
+  and serving reads), but not to do any deletions. This enables a multi-attached state for tenants,
+  facilitating live migration (this will be articulated in the next RFC that describes HA and
+  migrations).
+
+#### Visibility
+
+Because index_part.json is now written with a generation suffix, which data
+is visible depends on which generation the reader is operating in:
+
+- If two pageservers have the same tenant attached, they may have different
+  data visible as they're independently replaying the WAL, and maintaining
+  independent LayerMaps that are written to independent index_part.json files.
+  Data does not have to be remotely committed to be visible.
+- If one was passively reading from S3 from outside of a pageserver, the
+  visibility of data would depend on which index_part.json-<suffix> file
+  one had chosen to read from.
+- For a pageserver writing with a stale generation suffix, historic LSNs
+  remain readable until another pageserver (with a higher generation suffix)
+  decides to execute GC deletions. At this point, we may think of the stale
+  attachment's generation as having logically ended: during its existence
+  the generation had a consistent view of the world.
+- For a newly attached pageserver, its highest visible LSN may appears to
+  go backwards with respect to an earlier attachment, if that earlier
+  attachment had not uploaded all data to S3 before the new attachment.
+
+### Object Key Changes
+
+#### Generation suffix
+
+All object keys (layer objects and index objects) will contain a numeric suffix that
+advances as the attachment generation and node generation advance.
+This suffix is the primary mechanism for protecting against split-brain situations, and
+enabling safe multi-attachment of tenants:
+
+- Two pageservers running with the same node ID (e.g. after a failure, where there is
+  some rogue pageserver still running) will not try to write to the same objects.
+- Multiple attachments of the same tenant will not try to write to the same objects, as
+  each attachment would have a distinct attachment generation.
+  To avoid coupling our storage format tightly with the way we manage generation numbers,
+  they will be packed into a single u64 that must obey just the following properties:
```

Comment:
Indent

```suggestion
  each attachment would have a distinct attachment generation.

To avoid coupling our storage format tightly with the way we manage generation numbers,
they will be packed into a single u64 that must obey just the following properties:
```